packages/sats-wagmi/src/hooks/useBalance.tsx

```diff
@@ -29,10 +30,49 @@ const useBalance = (props: UseBalanceProps = {}) => {
       }
 
       const esploraClient = new EsploraClient(network);
+      const ordinalsClient = new OrdinalsClient(network);
 
-      const { confirmed, unconfirmed, total } = await esploraClient.getBalance(address);
+      const addressInfo = getAddressInfo(address);
 
-      return { confirmed: BigInt(confirmed), unconfirmed: BigInt(unconfirmed), total: BigInt(total) };
+      if (addressInfo.type === AddressType.p2tr) {
+        const [outputs, cardinalOutputs] = await Promise.all([
+          esploraClient.getAddressUtxos(address),
+          // cardinal = return UTXOs not containing inscriptions or runes
+          ordinalsClient.getOutputsFromAddress(address, 'cardinal')
+        ]);
+
+        const cardinalOutputsSet = new Set(cardinalOutputs.map((output) => output.outpoint));
+
+        const total = outputs.reduce((acc, output) => {
+          if (cardinalOutputsSet.has(OutPoint.toString(output))) {
+            return acc + output.value;
+          }
+
+          return acc;
+        }, 0);
+
+        const confirmed = outputs.reduce((acc, output) => {
+          if (cardinalOutputsSet.has(OutPoint.toString(output)) && output.confirmed) {
+            return acc + output.value;
+          }
+
+          return acc;
+        }, 0);
```

Comment:
Not: you are iterating over the outputs twice. Not sure what the complexity is of the JS under-the-hood but from a glance it would seem to me like it's better to check this once.

Reply:
time complexity wise it's `2*n` (or `n`)
space complexity is constant since we are computing number value